### PR TITLE
  Add feature "multicore"  and you can disable rayon by disabling the  "multicore" feature. 

### LINKED ISSUE #648: Ability to disable rayon
When compiling for wasm, we have to rely on [wasm-bindgen-rayon](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon#building-rust-code) which requires a Rust nightly.
We would like to stay on the Rust stable.
To that end, would it be possible to have a single-threaded mode in halo2 with rayon disabled?
Seems like the change is trivial and was already attempted earlier https://github.com/adria0/halo2/commit/e39100bb73d8122dd16fad191f4923f8e43f9b5a

If this single-threaded mode is ACKed, I could work on a PR.

### PR DESCRIPTION
**Pull request by nagatoism**
Close #648 All test has been passed with/without rayon.

### CODE REVIEW

**Review by CPerezz — COMMENTED** [review]
Looks good.

Only two comments:
1) I wonder if `cfg_if!` could abbreviate some of the repeated chunks of code here. Maybe worth looking at it.
2) I think this doesn't need to go through the `nightly` feature first. As it doesn't include anything new. But worth checking with @str4d .

For the rest, looks good!

File: README.md [hunk 1]
```diff
@@ -14,6 +14,9 @@ minor version bump.
 `halo2` currently uses [rayon](https://github.com/rayon-rs/rayon) for parallel computation.
 The `RAYON_NUM_THREADS` environment variable can be used to set the number of threads.
 
+You can disable [rayon] by disabling the ["multicore"] feature.
+Warning! Halo2 will lose access to parallelism if you disable the ["multicore"] feature. This will significantly downgrade performance.
```
Comment (daira) [inline README.md:18]:
```suggestion
Warning! Halo2 will lose access to parallelism if you disable the ["multicore"] feature.
This will significantly degrade performance.
```

File: halo2_proofs/Cargo.toml [hunk 2]
```diff
@@ -45,7 +45,7 @@ harness = false
 
 [dependencies]
 backtrace = { version = "0.3", optional = true }
-rayon = "1.5.1"
+rayon = {version = "1.5.1",optional = true}
```
Comment (daira) [inline halo2_proofs/Cargo.toml:48]:
```suggestion
rayon = {version = "1.5.1", optional = true}
```

File: halo2_proofs/Cargo.toml [hunk 3]
```diff
@@ -68,7 +68,8 @@ rand_core = { version = "0.6", default-features = false, features = ["getrandom"
 getrandom = { version = "0.2", features = ["js"] }
 
 [features]
-default = ["batch"]
+default = ["batch","multicore"]
```
Comment (daira) [inline halo2_proofs/Cargo.toml:71]:
```suggestion
default = ["batch", "multicore"]
```

File: halo2_proofs/README.md [hunk 4]
```diff
@@ -15,6 +15,9 @@ minor version bump.
 computation. The `RAYON_NUM_THREADS` environment variable can be used to set the number of
 threads.
 
+You can disable [rayon] by disabling the ["multicore"] feature.
+Warning! Halo2 will lose access to parallelism if you disable the ["multicore"] feature. This will significantly downgrade the performance.
```
Comment (daira) [inline halo2_proofs/README.md:19]:
```suggestion
Warning! Halo2 will lose access to parallelism if you disable the ["multicore"] feature.
This will significantly degrade the performance. 
```

File: halo2_proofs/src/arithmetic.rs [hunk 5]
```diff
@@ -152,19 +152,36 @@ pub fn best_multiexp<C: CurveAffine>(coeffs: &[C::Scalar], bases: &[C]) -> C::Cu
...
                 .zip(bases.chunks(chunk))
                 .zip(results.iter_mut())
             {
-                scope.spawn(move |_| {
-                    multiexp_serial(coeffs, bases, acc);
-                });
+                multiexp_serial(coeffs, bases, acc);
             }
-        });
+        };
```
Comment (daira) [inline halo2_proofs/src/arithmetic.rs:183]:
```suggestion
        }
```

File: halo2_proofs/src/arithmetic.rs [hunk 6]
```diff
@@ -262,10 +279,20 @@ pub fn recursive_butterfly_arithmetic<Scalar: Field, G: FftGroup<Scalar>>(
...
-        );
+
+        #[cfg(feature = "multicore")]
+        {
+            multicore::join(
+                || recursive_butterfly_arithmetic(left, n / 2, twiddle_chunk * 2, twiddles),
+                || recursive_butterfly_arithmetic(right, n / 2, twiddle_chunk * 2, twiddles),
+            );
+        }
+
```
Comment (daira) [inline halo2_proofs/src/arithmetic.rs:290]:
```suggestion
```

File: halo2_proofs/src/arithmetic.rs [hunk 7]
```diff
@@ -337,22 +364,30 @@ where
...
+        multicore::scope(|scope| {
+            for (chunk_num, v) in v.chunks_mut(chunk).enumerate() {
+                let f = f.clone();
+                scope.spawn(move |_| {
+                    let start = chunk_num * chunk;
+                    f(v, start);
+                });
+            }
+        });
     }
```
Comment (daira) [inline halo2_proofs/src/arithmetic.rs:386]:
```suggestion
```

File: halo2_proofs/src/multicore.rs [hunk 8]
```diff
@@ -1,9 +1,15 @@
...
 //! be extended in the future to allow for various parallelism strategies.
-
+#[cfg(feature = "multicore")]
 pub use rayon::{
     current_num_threads,
     iter::{IndexedParallelIterator, IntoParallelIterator, ParallelIterator},
     join, scope, Scope,
 };
+
+
```
Comment (daira) [inline halo2_proofs/src/multicore.rs:11]:
```suggestion
```

File: halo2_proofs/src/multicore.rs [hunk 9]
```diff
@@ -1,9 +1,15 @@
...
+#[cfg(not(feature = "multicore"))]
+pub fn current_num_threads() -> usize {
+    return 1;
```
Comment (daira) [inline halo2_proofs/src/multicore.rs:14]:
```suggestion
    1
```

File: halo2_proofs/src/plonk/verifier/batch.rs [hunk 10]
```diff
@@ -89,34 +92,70 @@ where
...
-                let instances: Vec<_> = instances.iter().map(|i| &i[..]).collect();
-
-                let strategy = BatchStrategy::new(params);
-                let mut transcript = Blake2bRead::init(&item.proof[..]);
-                verify_proof(params, vk, strategy, &instances, &mut transcript).map_err(|e| {
-                    tracing::debug!("Batch item {} failed verification: {}", i, e);
-                    e
+        #[cfg(feature = "multicore")]
+        {
+            let final_msm = self
```
Comment (daira) [inline halo2_proofs/src/plonk/verifier/batch.rs:97]:
```suggestion
        let final_msm = {
            self
```

File: halo2_proofs/src/plonk/verifier/batch.rs [hunk 11]
```diff
@@ -89,34 +92,70 @@ where
...
-            Err(_) => false,
+                .try_fold(
+                    || params.empty_msm(),
+                    |msm, res| res.map(|proof_msm| accumulate_msm(msm, proof_msm)),
+                )
+                .try_reduce(|| params.empty_msm(), |a, b| Ok(accumulate_msm(a, b)));
+            match final_msm {
+                Ok(msm) => msm.eval(),
+                Err(_) => false,
+            }
```
Comment (daira) [inline halo2_proofs/src/plonk/verifier/batch.rs:124]:
```suggestion
```
(moved to common code below)

File: halo2_proofs/src/plonk/verifier/batch.rs [hunk 12]
```diff
@@ -89,34 +92,70 @@ where
...
-            .try_reduce(|| params.empty_msm(), |a, b| Ok(accumulate_msm(a, b)));
-
-        match final_msm {
-            Ok(msm) => msm.eval(),
-            Err(_) => false,
+                .try_fold(
+                    || params.empty_msm(),
+                    |msm, res| res.map(|proof_msm| accumulate_msm(msm, proof_msm)),
+                )
+                .try_reduce(|| params.empty_msm(), |a, b| Ok(accumulate_msm(a, b)));
```
Comment (daira) [inline halo2_proofs/src/plonk/verifier/batch.rs:120]:
```suggestion
                .try_reduce(|| params.empty_msm(), |a, b| Ok(accumulate_msm(a, b)))
```

File: halo2_proofs/src/plonk/verifier/batch.rs [hunk 13]
```diff
@@ -89,34 +92,70 @@ where
...
+            match final_msm {
+                Ok(msm) => msm.eval(),
+                Err(_) => false,
+            }
+        }
+        #[cfg(not(feature = "multicore"))]
+        {
+            let final_msm = self
```
Comment (daira) [inline halo2_proofs/src/plonk/verifier/batch.rs:128]:
```suggestion
        let final_msm = {
            self
```

File: halo2_proofs/src/plonk/verifier/batch.rs [hunk 14]
```diff
@@ -89,34 +92,70 @@ where
...
+                    })
+                })
+                .try_fold(params.empty_msm(), |msm, res| {
+                    res.map(|proof_msm| accumulate_msm(msm, proof_msm))
+                });
+
+            let final_msm = match final_msm {
+                Ok(final_msm) => Ok(accumulate_msm(params.empty_msm(), final_msm)),
+                Err(err) => Err(err),
+            };
```
Comment (daira) [inline halo2_proofs/src/plonk/verifier/batch.rs:154]:
```suggestion
                })
                .map(|msm| accumulate_msm(params.empty_msm(), msm))
        }
```

File: halo2_proofs/src/plonk/verifier/batch.rs [hunk 15]
```diff
@@ -89,34 +92,70 @@ where
...
+            match final_msm {
+                Ok(msm) => msm.eval(),
+                Err(_) => false,
+            }
         }
```
Comment (daira) [inline halo2_proofs/src/plonk/verifier/batch.rs:159]:
```suggestion
        match final_msm {
            Ok(msm) => msm.eval(),
            Err(_) => false,
        }
```

File: halo2_proofs/src/poly/evaluator.rs [hunk 16]
```diff
@@ -210,20 +210,38 @@ impl<E, F: Field, B: Basis> Evaluator<E, F, B> {
...
+                            chunk_size,
+                            chunk_index,
+                            polys: &self.polys,
+                        };
+                        out.copy_from_slice(&recurse(ast, &ctx));
+                    });
+                }
+            });
+        }
+
```
Comment (daira) [inline halo2_proofs/src/poly/evaluator.rs:230]:
```suggestion
```

**Review by daira — APPROVED** [review]
utACK with suggested changes.

**Comment by str4d** [comment]
There is a lot of duplicated logic here, and the conditionals are spread out over the codebase. I'd prefer we take a similar approach to `bellman` and provide fake versions of the `rayon` types in the `halo2_proofs::multicore` module that just implement serial logic. Maybe we should move this "maybe rayon" logic into a crate?

And now that I go searching, someone else had the same idea a year ago: https://crates.io/crates/maybe-rayon

File: halo2_proofs/Cargo.toml (hunk 2 above)
Comment (str4d) [inline halo2_proofs/Cargo.toml:48]:
```suggestion
rayon = { package = "maybe-rayon", version = "0.1", default-features = false }
```

File: halo2_proofs/Cargo.toml [hunk 17]
```diff
@@ -68,7 +68,8 @@ rand_core = { version = "0.6", default-features = false, features = ["getrandom"
...
+multicore = ["rayon"]
```
Comment (str4d) [inline halo2_proofs/Cargo.toml:72]:
```suggestion
multicore = ["rayon/threads"]
```
(or `maybe-rayon/threads`, IDK if namespaced feature flags are resolved relative to the package name or the dependency name.)

File: halo2_proofs/src/arithmetic.rs [hunk 18]
```diff
@@ -152,19 +152,36 @@ pub fn best_multiexp<C: CurveAffine>(coeffs: &[C::Scalar], bases: &[C]) -> C::Cu
         let chunk = coeffs.len() / num_threads;
         let num_chunks = coeffs.chunks(chunk).len();
         let mut results = vec![C::Curve::identity(); num_chunks];
-        multicore::scope(|scope| {
+
+        #[cfg(feature = "multicore")]
```
Comment (str4d) [inline halo2_proofs/src/arithmetic.rs:156]: Undo these changes.

File: halo2_proofs/src/arithmetic.rs [hunk 19]
```diff
@@ -262,10 +279,20 @@ pub fn recursive_butterfly_arithmetic<Scalar: Field, G: FftGroup<Scalar>>(
         a[1] -= &t;
     } else {
         let (left, right) = a.split_at_mut(n / 2);
-        multicore::join(
-            || recursive_butterfly_arithmetic(left, n / 2, twiddle_chunk * 2, twiddles),
-            || recursive_butterfly_arithmetic(right, n / 2, twiddle_chunk * 2, twiddles),
-        );
+
+        #[cfg(feature = "multicore")]
```
Comment (str4d) [inline halo2_proofs/src/arithmetic.rs:283]: Undo these changes.

File: halo2_proofs/src/arithmetic.rs [hunk 20]
```diff
@@ -337,22 +364,30 @@ where
 /// This simple utility function will parallelize an operation that is to be
 /// performed over a mutable slice.
 pub fn parallelize<T: Send, F: Fn(&mut [T], usize) + Send + Sync + Clone>(v: &mut [T], f: F) {
-    let n = v.len();
-    let num_threads = multicore::current_num_threads();
-    let mut chunk = n / num_threads;
-    if chunk < num_threads {
-        chunk = n;
+    #[cfg(feature = "multicore")]
```
Comment (str4d) [inline halo2_proofs/src/arithmetic.rs:367]: Undo these changes.

File: halo2_proofs/src/multicore.rs [hunk 21]
```diff
@@ -1,9 +1,15 @@
 //! An interface for dealing with the kinds of parallel computations involved in
 //! `halo2`. It's currently just a (very!) thin wrapper around [`rayon`] but may
 //! be extended in the future to allow for various parallelism strategies.
-
+#[cfg(feature = "multicore")]
```
Comment (str4d) [inline halo2_proofs/src/multicore.rs:4]: Using `maybe-rayon`, I think this feature flag could just be removed and these imports left as unconditional (though we would need to keep `current_num_threads` separately conditional as I don't think `maybe-rayon` provides an equivalent).

File: halo2_proofs/src/plonk/verifier/batch.rs [hunk 22]
```diff
@@ -89,34 +92,70 @@ where
...
-                    .map(|i| i.iter().map(|c| &c[..]).collect())
-                    .collect();
-                let instances: Vec<_> = instances.iter().map(|i| &i[..]).collect();
-
-                let strategy = BatchStrategy::new(params);
-                let mut transcript = Blake2bRead::init(&item.proof[..]);
-                verify_proof(params, vk, strategy, &instances, &mut transcript).map_err(|e| {
-                    tracing::debug!("Batch item {} failed verification: {}", i, e);
-                    e
+        #[cfg(feature = "multicore")]
```
Comment (str4d) [inline halo2_proofs/src/plonk/verifier/batch.rs:95]: Undo these changes.

File: halo2_proofs/src/poly/evaluator.rs [hunk 23]
```diff
@@ -210,20 +210,38 @@ impl<E, F: Field, B: Basis> Evaluator<E, F, B> {
         // Apply `ast` to each chunk in parallel, writing the result into an output
         // polynomial.
         let mut result = B::empty_poly(domain);
-        multicore::scope(|scope| {
+        #[cfg(feature = "multicore")]
```
Comment (str4d) [inline halo2_proofs/src/poly/evaluator.rs:213]: Undo these changes.

**Review by str4d — CHANGES_REQUESTED** [review]
If we switch to using the https://crates.io/crates/maybe-rayon crate, I think that the only code changes in this PR should be to `halo2_proofs/src/multicore.rs`, which would greatly simplify it and make future maintenance simpler.

File: halo2_proofs/src/poly/evaluator.rs (hunk 16 above)
Comment (nagatoism) [reply halo2_proofs/src/poly/evaluator.rs:230]: Thanks for fixing the typos.  Will apply.

File: halo2_proofs/src/multicore.rs [hunk 24]
```diff
@@ -2,8 +2,15 @@
 //! `halo2`. It's currently just a (very!) thin wrapper around [`rayon`] but may
```
Comment (daira) [inline halo2_proofs/src/multicore.rs:2]:
```suggestion
//! `halo2`. It's currently just a (very!) thin wrapper around `rayon` but may
```
(This link broke because `rayon` is no longer a direct dependency.)

File: halo2_proofs/src/multicore.rs [hunk 25]
```diff
@@ -2,8 +2,15 @@
...
 //! be extended in the future to allow for various parallelism strategies.
 
-pub use rayon::{
-    current_num_threads,
+pub use maybe_rayon::{
     iter::{IndexedParallelIterator, IntoParallelIterator, ParallelIterator},
```
Comment (daira) [inline halo2_proofs/src/multicore.rs:6]: Please investigate the wasm32-wasi failure; I believe this was working on that platform before.

**Review by daira — CHANGES_REQUESTED** [review]
Please investigate the wasm32-wasi failures. Otherwise this is looking pretty good modulo str4d's comments.

**Comment by nagatoism** [comment]
Hi Found that IndexedParallelIter is not supported by maybe-rayon. The new code fixes this. 

The code is necessary because try_fold rayon and try_fold in rust std are different functions so I cannot reuse code. 
Here  https://docs.rs/rayon/latest/rayon/iter/trait.ParallelIterator.html#method.fold

PS pls help with the mut  on final_msm. No idea why the mut is needed.

**Comment by str4d** [comment]
If `IndexedParallelIter` isn't supported upstream, then can we just write our own implementation? That would avoid the try_fold issue.

**Comment by str4d** [comment]
CI is failing for an unrelated reason: `constant_time_eq` has raised its MSRV to 1.59. It is a dependency of `blake2b_simd`, so we will therefore need to similarly raise our MSRV (I'll do this in a separate PR).

File: README.md [hunk 26]
```diff
@@ -14,6 +14,10 @@ minor version bump.
 `halo2` currently uses [rayon](https://github.com/rayon-rs/rayon) for parallel computation.
 The `RAYON_NUM_THREADS` environment variable can be used to set the number of threads.
 
+You can disable [rayon] by disabling the ["multicore"] feature.
+Warning! Halo2 will lose access to parallelism if you disable the ["multicore"] feature.
```
Comment (daira) [inline README.md:18]:
```suggestion
You can disable `rayon` by disabling the `"multicore"` feature.
Warning! Halo2 will lose access to parallelism if you disable the `"multicore"` feature.
```

File: halo2_proofs/README.md [hunk 27]
```diff
@@ -15,6 +15,10 @@ minor version bump.
 computation. The `RAYON_NUM_THREADS` environment variable can be used to set the number of
 threads.
 
+You can disable [rayon] by disabling the ["multicore"] feature.
+Warning! Halo2 will lose access to parallelism if you disable the ["multicore"] feature.
```
Comment (daira) [inline halo2_proofs/README.md:19]:
```suggestion
You can disable `rayon` by disabling the `"multicore"` feature.
Warning! Halo2 will lose access to parallelism if you disable the `"multicore"` feature.
```

File: halo2_proofs/src/multicore.rs [hunk 28]
```diff
@@ -1,9 +1,61 @@
...
+#[cfg(not(feature = "multicore"))]
+pub trait IndexedParallelIterator: std::iter::Iterator {}
+
+pub trait TryFoldAndReduce<T, E> {
+    /// Implements `iter.try_fold().try_reduce()` for `rayon::iter::ParallelIterator`,
+    /// falling back on `Iterator::try_fold` when the `multicore` feature flag is
+    /// disabled.
+    /// try_fold_and_reduce function can only be called by a iter with
+    /// Result<T,E> item type because the folder_op Fn<T,T>->T must
+    /// meet both trait bound of rayon's try_fold and try_reduce```
```
Comment (daira) [inline halo2_proofs/src/multicore.rs:27]:
```suggestion
    /// The `try_fold_and_reduce` function can only be called by a iter with
    /// `Result<T, E>` item type because the `fold_op: Fn<T, T> -> T` must
    /// meet both the trait bounds of rayon's `try_fold` and `try_reduce`.   
```

File: halo2_proofs/src/multicore.rs [hunk 29]
```diff
@@ -1,9 +1,61 @@
...
+        fold_op: impl Fn(Result<T, E>, Result<T, E>) -> Result<T, E> + Send + Sync,
+    ) -> Result<T, E>;
+}
+
+#[cfg(feature = "multicore")]
+impl<
+        T: Send + Sync,
+        E: Send + Sync,
+        I: maybe_rayon::iter::ParallelIterator<Item = Result<T, E>>,
+    > TryFoldAndReduce<T, E> for I
```
Comment (daira) [inline halo2_proofs/src/multicore.rs:40]: I usually prefer to use `where` if the generic constraint would break across lines.

File: halo2_proofs/src/multicore.rs [hunk 30]
```diff
@@ -1,9 +1,61 @@
...
+    /// Implements `iter.try_fold().try_reduce()` for `rayon::iter::ParallelIterator`,
+    /// falling back on `Iterator::try_fold` when the `multicore` feature flag is
+    /// disabled.
+    /// try_fold_and_reduce function can only be called by a iter with
+    /// Result<T,E> item type because the folder_op Fn<T,T>->T must
+    /// meet both trait bound of rayon's try_fold and try_reduce```   
+    fn try_fold_and_reduce(
+        self,
+        identity: impl Fn() -> T + Send + Sync,
+        fold_op: impl Fn(Result<T, E>, Result<T, E>) -> Result<T, E> + Send + Sync,
```
Comment (daira) [inline halo2_proofs/src/multicore.rs:31]:
```suggestion
        fold_op: impl Fn(T, Result<T, E>) -> Result<T, E> + Send + Sync,
```
`fold_op` is always called with `Ok(...)` as its first argument in both the multicore and non-multicore implementations. This makes sense because the fold/reduce is only continued when the accumulator is ok.

File: halo2_proofs/src/multicore.rs [hunk 31]
```diff
@@ -1,9 +1,61 @@
...
+        I: maybe_rayon::iter::ParallelIterator<Item = Result<T, E>>,
+    > TryFoldAndReduce<T, E> for I
+{
+    fn try_fold_and_reduce(
+        self,
+        identity: impl Fn() -> T + Send + Sync,
+        fold_op: impl Fn(Result<T, E>, Result<T, E>) -> Result<T, E> + Send + Sync,
+    ) -> Result<T, E> {
+        self.try_fold(&identity, |msm, res| fold_op(Ok(msm), res))
+            .try_reduce(&identity, |a, b| fold_op(Ok(a), Ok(b)))
```
Comment (daira) [inline halo2_proofs/src/multicore.rs:48]:
```suggestion
        fold_op: impl Fn(T, Result<T, E>) -> Result<T, E> + Send + Sync,
    ) -> Result<T, E> {
        self.try_fold(&identity, fold_op).try_reduce(&identity, |a, b| fold_op(a, Ok(b)))
```

File: halo2_proofs/src/multicore.rs [hunk 32]
```diff
@@ -1,9 +1,61 @@
...
+}
+
+#[cfg(not(feature = "multicore"))]
+impl<T, E, I: std::iter::Iterator<Item = Result<T, E>>> TryFoldAndReduce<T, E> for I {
+    fn try_fold_and_reduce(
+        mut self,
+        identity: impl Fn() -> T + Send + Sync,
+        fold_op: impl Fn(Result<T, E>, Result<T, E>) -> Result<T, E> + Send + Sync,
+    ) -> Result<T, E> {
+        self.try_fold(identity(), |acc, res| fold_op(Ok(acc), res))
```
Comment (daira) [inline halo2_proofs/src/multicore.rs:59]:
```suggestion
        fold_op: impl Fn(T, Result<T, E>) -> Result<T, E> + Send + Sync,
    ) -> Result<T, E> {
        self.try_fold(identity(), fold_op)
```

File: halo2_proofs/src/plonk/verifier/batch.rs [hunk 33]
```diff
@@ -108,11 +111,10 @@ where
                     e
                 })
             })
-            .try_fold(
+            .try_fold_and_reduce(
                 || params.empty_msm(),
-                |msm, res| res.map(|proof_msm| accumulate_msm(msm, proof_msm)),
-            )
-            .try_reduce(|| params.empty_msm(), |a, b| Ok(accumulate_msm(a, b)));
+                |msm, res| msm.and_then(|acc| res.map(|proof_msm| accumulate_msm(acc, proof_msm))),
```
Comment (daira) [inline halo2_proofs/src/plonk/verifier/batch.rs:116]:
```suggestion
                |acc, res| res.map(|proof_msm| accumulate_msm(acc, proof_msm)),
```

**Review by daira — APPROVED** [review]
utACK with a proposed refactoring.

**Comment by str4d** [comment]
- [Force-pushed](https://github.com/zcash/halo2/compare/5d391f4584fac0c7574253d321f64b45ab26cde2..c9144b347c0beeae954fb9cc9718bbb5ceac73c1) to fix `rustfmt` and referencing errors after accepting a suggestion.
- [Rebased on `main`](https://github.com/zcash/halo2/compare/c9144b347c0beeae954fb9cc9718bbb5ceac73c1..b29306c6dea7bb6e30513d6b18b9124204293f75) to fix MSRV issue.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/halo2/compare/b29306c6dea7bb6e30513d6b18b9124204293f75..a6eef9e3cb0fb4fd480ff22717ef679f0d8b4100) to address @daira's comment.

**Comment by str4d** [comment]
CI is currently failing because of an unrelated issue. Fixing that in #734, and then I'll rebase this PR so we can see it fail in the expected way :)

**Comment by str4d** [comment]
[Rebased on `main`](https://github.com/zcash/halo2/compare/47e99978f5b100f95d43081df3a6a2a42814c6d3..2c9c105dc398e2662d63750cbc60b5c4ce31c585) to include #734.

**Comment by str4d** [comment]
And after [another force-push to fix my last commit](https://github.com/zcash/halo2/compare/2c9c105dc398e2662d63750cbc60b5c4ce31c585..f95e495951c79e3fe88df7fdbef88ab69da70275), the WASM CI test is failing as expected: in addition to the expected compilation issues (due to `maybe-rayon` falling back to single-threading for non-atomics WASM if `threads` is enabled) we now see our compiler error that `multicore` is not supported on non-atomics WASM.

**Review by str4d — APPROVED** [review]
ACK. Thanks for the PR @nagatoism!
